flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java

```diff
@@ -0,0 +1,214 @@
+     *     if specified. The system classpath is used if not specified.
+     * @param jarFile The jar archive used expecting to have the job class included.
+     * @param jobClassName The job class that will be used if specified. The classpath is used to
+     *     detect any main class if not specified.
+     * @param programArgs The program arguments.
+     * @param configuration The Flink configuration for the given job.
+     * @return The {@code PackageProgramRetrieverImpl} that can be used to create a {@link
+     *     PackagedProgram} instance.
+     * @throws FlinkException If something goes wrong during instantiation.
+     */
+    public static PackagedProgramRetrieverImpl create(
+            @Nullable File userLibDir,
+            @Nullable File jarFile,
+            @Nullable String jobClassName,
+            String[] programArgs,
+            Configuration configuration)
+            throws FlinkException {
+        List<URL> userClasspath;
+        try {
+            userClasspath = extractUserClasspath(userLibDir);
+        } catch (IOException e) {
+            throw new FlinkException("An error occurred while extracting the user classpath.", e);
+        }
+
+        final EntryClassInformationProvider entryClassInformationProvider =
+                createEntryClassInformationProvider(
+                        userLibDir, userClasspath, jarFile, jobClassName, programArgs);
+        return new PackagedProgramRetrieverImpl(
+                entryClassInformationProvider, programArgs, userClasspath, configuration);
+    }
+
+    @VisibleForTesting
+    static EntryClassInformationProvider createEntryClassInformationProvider(
+            @Nullable File userLibDir,
+            Iterable<URL> userClasspath,
+            @Nullable File jarFile,
+            @Nullable String jobClassName,
+            String[] programArgs)
+            throws FlinkException {
+        if (PackagedProgramUtils.isPython(jobClassName)
+                || PackagedProgramUtils.isPython(programArgs)) {
+            return FromJarEntryClassInformationProvider.createFromPythonJar();
+        }
+
+        if (jarFile != null) {
+            return FromJarEntryClassInformationProvider.createFromCustomJar(jarFile, jobClassName);
+        }
+
+        if (userLibDir != null) {
+            return fromUserClasspath(jobClassName, userClasspath);
+        }
+
+        return fromSystemClasspath(jobClassName);
+    }
+
+    static EntryClassInformationProvider fromSystemClasspath(@Nullable String jobClassName)
+            throws FlinkException {
+        if (jobClassName != null) {
+            return FromClasspathEntryClassInformationProvider
+                    .createWithJobClassAssumingOnSystemClasspath(jobClassName);
+        }
+
+        try {
+            return FromClasspathEntryClassInformationProvider.createFromSystemClasspath();
+        } catch (Throwable t) {
```

Comment:
Why do you catch Throwables? Why is catching FlinkExceptions not sufficient?

Reply:
You're right. Initialliy, we were only expecting `IOException`, `NoSuchElementException`, and `IllegalArgumentException`. I, initially, thought that it would be good to have all errors covered. But I make it more explicit now. Additionally, I transform the `NoSuchElementException` and `IllegalArgumentException` into a checked `FlinkException` within the `FromClasspathEntryClassInformationProvider` to make these two error cases more prominent in the interface.

Reply:
You could've just catched `Exception` to cover these cases.